middle_end/flambda2/from_lambda/lambda_to_flambda_primitives_helpers.ml

```diff
@@ -286,32 +342,58 @@ let rec bind_rec acc exn_cont ~register_const0 (prim : expr_primitive)
         (Bound_pattern.singleton cond_result_pat)
         cond ~body:switch
     in
-    let join_handler_expr acc =
-      cont acc (Named.create_simple (Simple.var result_var))
-    in
-    let ifso_handler_expr acc =
-      bind_rec acc exn_cont ~register_const0 ifso dbg @@ fun acc ifso ->
+    let join_handler_expr acc = cont acc result_nameds in
+    let ifso_handler_expr acc : Expr_with_acc.t =
+      bind_recs acc exn_cont ~register_const0 ifso dbg @@ fun acc ifso ->
+      let ifso_result_vars =
+        List.map (fun _ -> Variable.create "ifso_result") ifso
+      in
+      let ifso_result_pats =
+        List.map
+          (fun ifso_result_var ->
+            Bound_var.create ifso_result_var Name_mode.normal)
+          ifso_result_vars
+      in
+      let ifso_result_simples = List.map Simple.var ifso_result_vars in
       let acc, apply_cont =
-        Apply_cont_with_acc.create acc join_point_cont
-          ~args:[Simple.var ifso_result]
+        Apply_cont_with_acc.create acc join_point_cont ~args:ifso_result_simples
           ~dbg
       in
       let acc, body = Expr_with_acc.create_apply_cont acc apply_cont in
-      Let_with_acc.create acc
-        (Bound_pattern.singleton ifso_result_pat)
-        ifso ~body
+      List.fold_left2
+        (fun (acc, body) ifso_result_pat ifso ->
+          Let_with_acc.create acc
+            (Bound_pattern.singleton ifso_result_pat)
+            ifso ~body)
+        (acc, body)
+        (List.rev ifso_result_pats)
+        (List.rev ifso)
     in
-    let ifnot_handler_expr acc =
-      bind_rec acc exn_cont ~register_const0 ifnot dbg @@ fun acc ifnot ->
+    let ifnot_handler_expr acc : Expr_with_acc.t =
+      bind_recs acc exn_cont ~register_const0 ifnot dbg @@ fun acc ifnot ->
+      let ifnot_result_vars =
+        List.map (fun _ -> Variable.create "ifnot_result") ifnot
+      in
+      let ifnot_result_pats =
+        List.map
+          (fun ifnot_result_var ->
+            Bound_var.create ifnot_result_var Name_mode.normal)
+          ifnot_result_vars
+      in
+      let ifnot_result_simples = List.map Simple.var ifnot_result_vars in
       let acc, apply_cont =
         Apply_cont_with_acc.create acc join_point_cont
-          ~args:[Simple.var ifnot_result]
-          ~dbg
+          ~args:ifnot_result_simples ~dbg
       in
       let acc, body = Expr_with_acc.create_apply_cont acc apply_cont in
-      Let_with_acc.create acc
-        (Bound_pattern.singleton ifnot_result_pat)
-        ifnot ~body
+      List.fold_left2
+        (fun (acc, body) ifnot_result_pat ifnot ->
+          Let_with_acc.create acc
+            (Bound_pattern.singleton ifnot_result_pat)
+            ifnot ~body)
+        (acc, body)
+        (List.rev ifnot_result_pats)
+        (List.rev ifnot)
```

Comment:
All this code is duplicated in the ifso/ifnot cases, could we share the code a bit to simplify it?